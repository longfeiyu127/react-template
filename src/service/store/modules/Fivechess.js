// 1.横向
function Crosswise(squares, i, j) {
  // 前面的相同棋子
  const target = squares[i][j];
  let [prev, next] = [0, 0];
  while (true) {
    if (j - prev - 1 < 0 || squares[i][j - prev - 1] !== target) {
      break;
    }
    prev++;
  }
  while (true) {
    if (j + next + 1 > 14 || squares[i][j + next + 1] !== target) {
      break;
    }
    next++;
  }
  return prev + next === 4 ? [[i, j - prev], [i, j + next]] : false;
}
// 2.竖向
function Vertical(squares, i, j) {
  const target = squares[i][j];
  let [prev, next] = [0, 0];
  while (true) {
    if (i - prev - 1 < 0 || squares[i - prev - 1][j] !== target) {
      break;
    }
    prev++;
  }
  while (true) {
    if (i + next + 1 > 14 || squares[i + next + 1][j] !== target) {
      break;
    }
    next++;
  }
  return prev + next === 4 ? [[i - prev, j], [i + next, j]] : false;
}
// 3.左上至右下
function TopLeft(squares, i, j) {
  const target = squares[i][j];
  let [prev, next] = [0, 0];
  while (true) {
    if (
      i - prev - 1 < 0 ||
      j - prev - 1 < 0 ||
      squares[i - prev - 1][j - prev - 1] !== target
    ) {
      break;
    }
    prev++;
  }
  while (true) {
    if (
      i + next + 1 > 14 ||
      j + next + 1 > 14 ||
      squares[i + next + 1][j + next + 1] !== target
    ) {
      break;
    }
    next++;
  }
  return prev + next === 4
    ? [[i - prev, j - prev], [i + next, j + next]]
    : false;
}
// 4.右上至左下
function TopRight(squares, i, j) {
  const target = squares[i][j];
  let [prev, next] = [0, 0];
  while (true) {
    if (
      i - prev - 1 < 0 ||
      j + prev + 1 > 14 ||
      squares[i - prev - 1][j + prev + 1] !== target
    ) {
      break;
    }
    prev++;
  }
  while (true) {
    if (
      i + next + 1 > 14 ||
      j - next - 1 < 0 ||
      squares[i + next + 1][j - next - 1] !== target
    ) {
      break;
    }
    next++;
  }
  return prev + next === 4
    ? [[i - prev, j + prev], [i + next, j - next]]
    : false;
}

function computeCoord(startPoint, endPoint) {
  const [startI, startJ] = startPoint;
  const stepI = (endPoint[0] - startPoint[0]) / 4;
  const stepJ = (endPoint[1] - startPoint[1]) / 4;
  const result = new Array(5).fill(undefined);
  return result.map((item, index) => [
    startI + index * stepI,
    startJ + index * stepJ
  ]);
}

function calculateWinner(...data) {
  const [squares, i, j] = data;
  const piece = squares[i][j];
  const result =
    Crosswise(...data) ||
    Vertical(...data) ||
    TopLeft(...data) ||
    TopRight(...data) ||
    undefined;
  if (result) {
    return {
      winnerCoord: computeCoord(...result),
      winner: piece
    };
  }
  return result;
}

const Fivechess = {
  state: {
    history: [
      {
        squares: new Array(15).fill(new Array(15).fill(null))
      }
    ],
    isBlack: true, // 'Black' 'White',
    winner: undefined,
    winnerCoord: null
  },
  reducers: {
    setChess(state, { i, j }) {
      const { history, isBlack, winner } = state;
      const current = history[history.length - 1];
      if (current.squares[i][j] || winner) {
        return {
          ...state
        };
      }
      const squares = JSON.parse(JSON.stringify(current.squares));
      squares[i][j] = isBlack ? 'Black' : 'White';
      const calWinner = calculateWinner(squares, i, j);
      let data = {
        history: history.concat([
          {
            squares
          }
        ]),
        isBlack: !isBlack
      };
      if (calWinner) {
        data = {
          ...data,
          ...calWinner
        };
      }
      return {
        ...state,
        ...data
      };
    },
    setReset(state) {
      return {
        ...state,
        history: [
          {
            squares: new Array(15).fill(new Array(15).fill(null))
          }
        ],
        isBlack: true, // 'Black' 'White',
        winner: undefined,
        winnerCoord: null
      };
    }
  },
  effects: {}
};

export default Fivechess;
