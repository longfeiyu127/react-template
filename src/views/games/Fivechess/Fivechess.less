.g-fivechess{
  height: 100%;
  background-color: #a39d8d;
  // background: url('../../../assets//images/games/fivechess/chessboard.jpg') no-repeat;
  // background-size: 100% 100%;
}