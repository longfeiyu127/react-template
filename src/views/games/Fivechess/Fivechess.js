import React from 'react';
import { connect } from 'react-redux';
import Header from '../../../components/games/Fivechess/Header/Header';
import Board from '../../../components/games/Fivechess/Board/Board';
import Footer from '../../../components/games/Fivechess/Footer/Footer';
import Victory from '../../../components/games/Fivechess/Victory/Victory';
import Firework from '../../../components/common/Firework/Firework';
// // import Scoreboard from './components/Scoreboard'
import './Fivechess.less';
import { fullScreen, isFullscreen } from '../../../utils/fullScreen';

function Fivechess(props) {
  // eslint-disable-next-line
  const { history, winner, winnerCoord, isBlack, handleClick, reset } = props;
  const current = history[history.length - 1];
  return (
    <div className="g-fivechess">
      <Header
        isBlack={isBlack}
        winner={winner}
        isFullscreen={isFullscreen}
        fullScreen={fullScreen}
      />
      <Board
        squares={current.squares}
        winner={winner}
        winnerCoord={winnerCoord}
        onClick={(i, j) => handleClick(i, j)}
      />
      <Footer />
      <Firework gameover={winner}>
        <Victory winner={winner} reset={() => reset()} />
      </Firework>
    </div>
  );
}

const mapStateToProps = state => ({
  history: state.Fivechess.history,
  isBlack: state.Fivechess.isBlack,
  winner: state.Fivechess.winner,
  winnerCoord: state.Fivechess.winnerCoord
});

const mapDispatchToProps = dispatch => ({
  handleClick: dispatch.Fivechess.setChess,
  reset: dispatch.Fivechess.setReset
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Fivechess);
