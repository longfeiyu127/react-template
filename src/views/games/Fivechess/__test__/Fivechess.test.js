import React from 'react';
import renderer from 'react-test-renderer';
import { Fivechess } from '../Fivechess';

describe('<Fivechess />', () => {
  it('component render', () => {
    const history = [{ squares: [] }];

    const winner = 'whitle';

    const winnerCoord = '';

    const isBlack = false;

    const handleClick = () => {};

    const reset = '';
    let footer = renderer
      .create(
        <Fivechess
          history={history}
          winner={winner}
          winnerCoord={winnerCoord}
          isBlack={isBlack}
          handleClick={handleClick}
          reset={reset}
        />
      )
      .toJSON();
    expect(footer).toMatchSnapshot();

    footer = renderer
      .create(
        <Fivechess
          history={history}
          winner={winner}
          winnerCoord={winnerCoord}
          isBlack={isBlack}
          handleClick={handleClick}
          reset={reset}
        />
      )
      .toJSON();
    expect(footer).toMatchSnapshot();
  });
});
