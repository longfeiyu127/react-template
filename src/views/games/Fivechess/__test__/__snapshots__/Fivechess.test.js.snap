// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fivechess /> component render 1`] = `
<div
  className="g-fivechess"
>
  <header
    className="g-fic-header"
  >
    <div
      className="maximize iconfont icon-maximize"
      name="fullScreen"
      onClick={[Function]}
    />
    <div
      className="player"
    >
      <div
        className="spark hide"
      />
      <div
        className="piece pieceWhite"
      />
    </div>
  </header>
  <div
    className="g-fic-board"
  >
    <main
      className="board-main"
    />
    <footer
      className="board-footer"
    />
  </div>
  <footer
    className="g-fic-footer"
  />
  <div
    className="firework-box "
  >
    <div
      className="firework"
    >
      <div
        className="before"
      />
      <div
        className="after"
      />
    </div>
    <div
      className="g-fic-victory"
    >
      <p
        className="victory"
      >
        victory： 
        whitle
      </p>
      <p
        className="result"
        onClick={[Function]}
      >
        Again
      </p>
    </div>
  </div>
</div>
`;

exports[`<Fivechess /> component render 2`] = `
<div
  className="g-fivechess"
>
  <header
    className="g-fic-header"
  >
    <div
      className="maximize iconfont icon-maximize"
      name="fullScreen"
      onClick={[Function]}
    />
    <div
      className="player"
    >
      <div
        className="spark hide"
      />
      <div
        className="piece pieceWhite"
      />
    </div>
  </header>
  <div
    className="g-fic-board"
  >
    <main
      className="board-main"
    />
    <footer
      className="board-footer"
    />
  </div>
  <footer
    className="g-fic-footer"
  />
  <div
    className="firework-box "
  >
    <div
      className="firework"
    >
      <div
        className="before"
      />
      <div
        className="after"
      />
    </div>
    <div
      className="g-fic-victory"
    >
      <p
        className="victory"
      >
        victory： 
        whitle
      </p>
      <p
        className="result"
        onClick={[Function]}
      >
        Again
      </p>
    </div>
  </div>
</div>
`;
